### ISSUE TITLE
Backwards compatibility policy

### ISSUE THREAD

**Issue by brentyi** [nerfstudio-project/nerfstudio#2458]
It would be nice to have an explicit backwards compatibility policy.

Some suggestions:
- Everything must be backwards compatible
- Semantic versioning
- Some window of deprecation warnings

I'm really okay with anything. They all seem unideal, but my feeling is something just needs to be written down and adhered to; the alternative is leaving a lot of PRs (#2092, #2450, ...) in purgatory + avoiding anything but minor changes to the repo. This feels unsustainable.

An alternative model is to move toward an approach like [cleanrl](https://github.com/vwxyzjn/cleanrl), where core algorithmic details are designed to be copy-and-pasted, duplicated, etc. I like this but probably the code would need to be architected very differently to make the experience nice. 🙂 

cc @vye16

**Comment by Mxbonn** [comment]
Thanks for addressing this, I agree that a policy for this is needed as new approaches to NeRF may require substantial changes to the code base in the future.
Personally, I prefer the semantic versioning + deprecation window approach.

And while I appreciate the cleanrl approach, I think this code base has been going in the exact opposite direction with all its abstractions that converting this code base to a non-modular approach would take longer than creating a cleannerf repo from scratch.

**Comment by Ilyabasharov** [comment]
Fully agree with last paragraph of @brentyi. I think this framework under active development right now and did not support lots of functionality (gaussian, as easy example). After supporting this, can go to backwards compatibility. You can define a list of features that should be completed before this and move on to a new stage of versioning.

**Comment by pwais** [comment]
The new-ish Nerfstudio template ( https://github.com/nerfstudio-project/nerfstudio-method-template ) is a pretty decent "API" for isolating core from research space.  Could require any new algos to start there and then "graduate" to batteries-included Nerfstudio, and then try to support backwards-compat for _only_ the nerfstudio core versus particular algos, which may include nerfacto.  

The template could also be expanded a bit to include:
 * API for data processing pipelines / `ns-process-data`.  
 * API for offline rendering; `_render_trajectory_video()` is doing a lot of work...

There are of course things like the cited camera optimization PR that are pretty sweeping and don't really fit the template.  BUT the automated testing for nerfstudio today is a bit lacking; there are several changes that got merged and then broke lots of stuff and required follow-up PRs.  So today, it's easy to end up with a sweeping PR that only really works on your own computer.

So for larger changes, one strategy could be to force such changes to use the template, and to even copy-pasta core code as needed, then "graduate" certain things to core nerfstudio.  E.g. a 'contrib module' approach, where backwards compatibility and other expectations are relaxed prior to graduation.  Nerfstudio-the-project would probably want at least one full-time maintainer to help shepherd contrib->core, and it's totally possible for that to be too much to ask for at this point of the project.


For the "version-managed core," a key problem for backwards compatibility is schema evolution of the config files, transforms, camera path, and maybe even some checkpoints.  But could easily embrace a lightweight protobuf-like convention:
 * Add a field? Throw an error if it's required and missing
 * Never change the type of a field; only create new fields.
 * Delete a field?  Comment it out (don't delete it, leave breadcrumbs) and write some sort of deprecation warning if some code ever reads it.

For a pure-python, no-protobuf impl, stuff like `pydantic` offers a lot of the boilerplate validation built-in.  Maybe tyro already has some of that?

For weights, my vote is to assume NO backwards compatibility, and that the git commit + nerfstudio version info should be baked into checkpoints.  Seems par for the course in ML that you have to either re-train or do your own weight surgery when the code changes substantially.  But that can certainly be a foreign concept to e.g. 3D artists.


A good case study could be taking a bit of time to estimate what it would take to merge SDFStudio back into Nerfstudio, and think about where backwards compatibility plays in there.  While I doubt such a merge would ever happen, it's a pretty good example of the copy-pasta lifecycle for this repo.

**Comment by pwais** [comment]
Well if there’s limited support for backwards compatibility, would be cool
if at least the configs and torch checkpoints all get stamped with the
nerfstudio version + git commit if available.   That makes archeology
easier and can help anybody who opts in to implementing backwards support
to figure out if the user is trying to do something deprecated etc.



>

**Comment by brentyi** [comment]
Makes sense to me! Perhaps we can just write a file with information like a commit hash or version number in the outputs directory where the checkpoints/config files are serialized.
